File: front/src/api/clip_annotations.ts
import { z } from "zod";
import { AxiosInstance } from "axios";

import {
  type Clip,
  type Tag,
  type Note,
  type ClipAnnotation,
  ClipAnnotationSchema,
} from "@/api/schemas";
import { type NoteCreate, NoteCreateSchema } from "@/api/notes";
import { GetManySchema, Page } from "@/api/common";

export const ClipAnnotationPageSchema = Page(ClipAnnotationSchema);

export type ClipAnnotationPage = z.infer<typeof ClipAnnotationPageSchema>;

export const ClipAnnotationFilterSchema = z.object({
  clip__eq: z.string().uuid().optional(),
  tag__key: z.string().optional(),
  tag__value: z.string().optional(),
  annotation_project__eq: z.string().uuid().optional(),
  evaluation_set__eq: z.string().uuid().optional(),
});

export type ClipAnnotationFilter = z.input<typeof ClipAnnotationFilterSchema>;

export const GetClipAnnotationsQuerySchema = z.intersection(
  GetManySchema,
  ClipAnnotationFilterSchema,
);

export type GetClipAnnotationsQuery = z.input<
  typeof GetClipAnnotationsQuerySchema
>;

const DEFAULT_ENDPOINTS = {
  getMany: "/api/v1/clip_annotations/",
  get: "/api/v1/clip_annotations/detail/",
  create: "/api/v1/clip_annotations/",
  delete: "/api/v1/clip_annotations/",
  addTag: "/api/v1/clip_annotations/detail/tags/",
  removeTag: "/api/v1/clip_annotations/detail/tags/",
  addNote: "/api/v1/clip_annotations/detail/notes/",
  removeNote: "/api/v1/clip_annotations/detail/notes/",
};

export function registerClipAnnotationsAPI(
  instance: AxiosInstance,
  endpoints: typeof DEFAULT_ENDPOINTS = DEFAULT_ENDPOINTS,
) {
  async function create(clip: Clip): Promise<ClipAnnotation> {
    const response = await instance.post(
      endpoints.create,
      {},
      {
        params: { clip_uuid: clip.uuid },
      },
    );
    return ClipAnnotationSchema.parse(response.data);
  }

  async function getMany(
    query: GetClipAnnotationsQuery,
  ): Promise<ClipAnnotationPage> {
    const params = GetClipAnnotationsQuerySchema.parse(query);
    const response = await instance.get(endpoints.getMany, { params });
    return ClipAnnotationPageSchema.parse(response.data);
  }

  async function get(uuid: string): Promise<ClipAnnotation> {
    const response = await instance.get(endpoints.get, {
      params: { clip_annotation_uuid: uuid },
    });
    return ClipAnnotationSchema.parse(response.data);
  }

  async function deleteClipAnnotation(
    clipAnnotation: ClipAnnotation,
  ): Promise<ClipAnnotation> {
    const response = await instance.delete(endpoints.delete, {
      params: { clip_annotation_uuid: clipAnnotation.uuid },
    });
    return ClipAnnotationSchema.parse(response.data);
  }

  async function addTag(
    clipAnnotation: ClipAnnotation,
    tag: Tag,
  ): Promise<ClipAnnotation> {
    const respose = await instance.post(
      endpoints.addTag,
      {},
      {
        params: {
          clip_annotation_uuid: clipAnnotation.uuid,
          key: tag.key,
          value: tag.value,
        },
      },
    );
    return ClipAnnotationSchema.parse(respose.data);
  }

  async function removeTag(
    clipAnnotation: ClipAnnotation,
    tag: Tag,
  ): Promise<ClipAnnotation> {
    const respose = await instance.delete(endpoints.removeTag, {
      params: {
        clip_annotation_uuid: clipAnnotation.uuid,
        key: tag.key,
        value: tag.value,
      },
    });
    return ClipAnnotationSchema.parse(respose.data);
  }

  async function addNote(
    clipAnnotation: ClipAnnotation,
    data: NoteCreate,
  ): Promise<ClipAnnotation> {
    const body = NoteCreateSchema.parse(data);
    const response = await instance.post(endpoints.addNote, body, {
      params: {
        clip_annotation_uuid: clipAnnotation.uuid,
      },
    });
    return ClipAnnotationSchema.parse(response.data);
  }

  async function removeNote(
    clipAnnotation: ClipAnnotation,
    note: Note,
  ): Promise<ClipAnnotation> {
    const response = await instance.delete(endpoints.removeNote, {
      params: {
        clip_annotation_uuid: clipAnnotation.uuid,
        note_uuid: note.uuid,
      },
    });
    return ClipAnnotationSchema.parse(response.data);
  }

  return {
    create,
    getMany,
    get,
    delete: deleteClipAnnotation,
    addTag,
    removeTag,
    addNote,
    removeNote,
  } as const;
}

File: back/src/whombat/routes/clip_annotations.py
"""REST API routes for clip_annotations."""

from uuid import UUID

from fastapi import APIRouter, Depends

from whombat import api, schemas
from whombat.filters.clip_annotations import ClipAnnotationFilter
from whombat.routes.dependencies import Session, get_current_user_dependency
from whombat.routes.dependencies.settings import WhombatSettings
from whombat.routes.types import Limit, Offset

__all__ = [
    "get_clip_annotations_router",
]


def get_clip_annotations_router(settings: WhombatSettings) -> APIRouter:
    """Get the API router for clip_annotations."""

    active_user = get_current_user_dependency(settings)

    clip_annotations_router = APIRouter()

    @clip_annotations_router.post(
        "/",
        response_model=schemas.ClipAnnotation,
    )
    async def create_annotation(
        session: Session,
        clip_uuid: UUID,
    ):
        """Create annotation."""
        clip = await api.clips.get(session, clip_uuid)
        clip_annotation = await api.clip_annotations.create(
            session,
            clip=clip,
        )
        await session.commit()
        return clip_annotation

    @clip_annotations_router.get(
        "/",
        response_model=schemas.Page[schemas.ClipAnnotation],
    )
    async def get_clip_annotations(
        session: Session,
        limit: Limit = 10,
        offset: Offset = 0,
        filter: ClipAnnotationFilter = Depends(ClipAnnotationFilter),  # type: ignore
        sort_by: str = "-created_on",
    ):
        """Get a page of annotation clip_annotations."""
        (
            clip_annotations,
            total,
        ) = await api.clip_annotations.get_many(
            session,
            limit=limit,
            offset=offset,
            filters=[filter],
            sort_by=sort_by,
        )
        return schemas.Page(
            items=clip_annotations,
            total=total,
            limit=limit,
            offset=offset,
        )

    @clip_annotations_router.get(
        "/detail/",
        response_model=schemas.ClipAnnotation,
    )
    async def get_annotation(
        session: Session,
        clip_annotation_uuid: UUID,
    ):
        """Get an annotation annotation."""
        clip_annotation = await api.clip_annotations.get(
            session,
            clip_annotation_uuid,
        )
        return clip_annotation

    @clip_annotations_router.delete(
        "/detail/",
        response_model=schemas.ClipAnnotation,
    )
    async def delete_annotation(
        session: Session,
        clip_annotation_uuid: UUID,
    ):
        """Remove a clip from an annotation project."""
        clip_annotation = await api.clip_annotations.get(
            session,
            clip_annotation_uuid,
        )
        clip_annotaiton = await api.clip_annotations.delete(
            session,
            clip_annotation,
        )
        await session.commit()
        return clip_annotaiton

    @clip_annotations_router.post(
        "/detail/tags/",
        response_model=schemas.ClipAnnotation,
    )
    async def add_annotation_tag(
        session: Session,
        clip_annotation_uuid: UUID,
        key: str,
        value: str,
        user: schemas.SimpleUser = Depends(active_user),
    ):
        """Add a tag to an annotation annotation."""
        clip_annotation = await api.clip_annotations.get(
            session,
            clip_annotation_uuid,
        )
        tag = await api.tags.get(session, (key, value))
        clip_annotation = await api.clip_annotations.add_tag(
            session,
            clip_annotation,
            tag,
            user,
        )
        await session.commit()
        return clip_annotation

    @clip_annotations_router.delete(
        "/detail/tags/",
        response_model=schemas.ClipAnnotation,
    )
    async def remove_annotation_tag(
        session: Session,
        clip_annotation_uuid: UUID,
        key: str,
        value: str,
    ):
        """Remove a tag from an annotation annotation."""
        clip_annotation = await api.clip_annotations.get(
            session,
            clip_annotation_uuid,
        )
        tag = await api.tags.get(session, (key, value))
        clip_annotation = await api.clip_annotations.remove_tag(
            session,
            clip_annotation,
            tag,
        )
        await session.commit()
        return clip_annotation

    @clip_annotations_router.post(
        "/detail/notes/",
        response_model=schemas.ClipAnnotation,
    )
    async def create_annotation_note(
        session: Session,
        clip_annotation_uuid: UUID,
        data: schemas.NoteCreate,
        user: schemas.SimpleUser = Depends(active_user),
    ):
        """Create a note for an annotation annotation."""
        clip_annotation = await api.clip_annotations.get(
            session,
            clip_annotation_uuid,
        )
        note = await api.notes.create(
            session,
            message=data.message,
            is_issue=data.is_issue,
            created_by=user,
        )
        clip_annotation = await api.clip_annotations.add_note(
            session,
            clip_annotation,
            note,
        )
        await session.commit()
        return clip_annotation

    @clip_annotations_router.delete(
        "/detail/notes/",
        response_model=schemas.ClipAnnotation,
    )
    async def delete_annotation_note(
        session: Session,
        clip_annotation_uuid: UUID,
        note_uuid: UUID,
    ):
        """Delete a note from an annotation annotation."""
        clip_annotation = await api.clip_annotations.get(
            session,
            clip_annotation_uuid,
        )
        note = await api.notes.get(session, note_uuid)
        clip_annotation = await api.clip_annotations.remove_note(
            session,
            clip_annotation,
            note,
        )
        await session.commit()
        return clip_annotation

    return clip_annotations_router
